Add driver for SinoWealth-based mice

staticssleever668 opened the pull request:
Continuing on #24. This does not yet work at all. I managed to get it working locally, but it involved some terrible hacky work-arounds and I want to make it sane before staging. I will give a heads-up once this is complete.

whot committed:
Add a completely untested version of the sinowealth driver

staticssleever668 committed:
sinowealth: put a magic number into a variable

staticssleever668 committed:
sinowealth: use all bits for report rate

whot commented on ratbag/drivers/sinowealth.py:
```diff
@@ -24,6 +24,7 @@
 CONFIG_SIZE_USED_MAX = 167
 
 CONFIG_FLAGS_XY_RESOLUTION = 0x80
+CONFIG_FLAGS_REPORT_RATE = 0x0F
```
maybe suffix this with `_MASK`?

whot commented on ratbag/drivers/sinowealth.py:
```diff
@@ -237,7 +237,7 @@ def start(self):
         # now set up the ratbag device
         p = ratbag.Profile(
-            device=self.ratbg_device,
+            device=self.ratbag_device,
```
/me weeps in a statically compiled language

staticssleever668 committed:
sinowealth: add support for reading enabled resolutions

staticssleever668 committed:
sinowealth: fix DPI range values

staticssleever668 committed:
sinowealth: always enable XY_INDEPENDENT capability

The byte in config data shows not whether it can be used, but whether
it's activated right now.

staticssleever668 committed:
sinowealth: fix first two bytes in queries being swallowed

Doesn't seem like these bytes were used anyway.

staticssleever668 committed:
sinowealth: add a TODO to support different sensors in DPI converter

whot commented on ratbag/drivers/sinowealth.py:
```diff
@@ -166,10 +166,7 @@ def run(self, rodent: ratbag.driver.Rodent) -> Reply:
-            Spec("B", "report_id"),
-            Spec("B", "cmd"),
-        ] + self.reply_spec
+        rspec = self.reply_spec
```
this seems strange - in the libratbag driver the first two fields of `struct sinowealth_config_report` are the `report_id` and `command_id`. So I'd expect these to be in place here? If you drop them from the `Spec` list your parsing will be off by 2 bytes now.

staticssleever668 replied:
They are already in the config report struct.
Before this change the spec was like this:
- Report ID
- CMD
- Report ID
- CMD
- Actual config report data..

whot replied:
oh, woops, sorry about the noise. This approach with a standard header is more interesting for drivers with lots of reports (hidpp20 for example) but here we have only 2 I think, so let's skip it.

You won't need `rspec` then btw, you can just pass `self.reply_spec` to the parser. but that's a nitpick only.

whot commented on ratbag/drivers/sinowealth.py:
```diff
@@ -72,8 +73,8 @@ def from_bytes(cls, data: bytes) -> "Config":
             Spec("B", "dpi_count"),  # two nibbles!
-            Spec("B", "dpi_enabled"),
             Spec("B", "dpi", repeat=16),
+            Spec("B", "dpi_disabled_slots"),  # bit mask
```
this doesn't match the `sinowealth_config_report` where we have dpi_count, then the dpi. Note that the `Spec` entries are parsed in-order, so one of this or libratbag seems off now.

the `Spec` approach doesn't allow for nibbles so config and dpi_count here both parse one byte that then gets split into report_rate and active_dpi as well.

staticssleever668 replied:
Yes, thank you. I renamed a bunch of fields but did not stage it, so things I did stage got a bit out of order.

staticssleever668 committed:
sinowealth: rename a bunch of fields

staticssleever668 committed:
sinowealth: add all known config data fields

staticssleever668 committed:
sinowealth: fix several bytes missing in firmware version query

staticssleever668 committed:
sinowealth: check if command in query reply is correct

staticssleever668 committed:
sinowealth: fix config queries

staticssleever668 committed:
sinowealth: fix config queries for long devices

staticssleever668 commented:
Rebased on master on squashed a few commits. New ones go after [sinowealth: rename a bunch of fields](https://github.com/whot/ratbag-python/pull/25/commits/2d88c20e2cfa8d195854d7b2b57a00c13da50ab9).
This now actually works. It displays correct DPI profiles (but doesn't show the active one yet) and shows correct report rate.

@whot I tried to add firmware version field into the ratbag.Device, but the code structure I got was kind of ugly: I moved ratbag.Device creation into `start` method of SinowealthDevice, and made it return the Device. Maybe you could guide me into a better way to do it?

whot commented:
fwiw, I purposely didn't put much of an API around the device initialization. There's the driver entry point `new_with_devicelist` but since we're using the `HidrawDriver` here for some of the boilerplate the only API we need is: at some point `probe()` is called by ratbag. And following that, you should either throw an exception if the device is not supported/missing something. Or (eventually) emit the `device-added` signal (one per device, you could have multiple per hidraw device). Beyond that, anything else is left to the driver to do as they see fit, even the signal doesn't need to be emitted in immediate response to `start`, you could schedule some work for the future which then emits the signal.

So this is the long way to say "whatever works for you" :)

The other two drivers already have the ratbag.Device creation inside `start()` so that's definitely a good option. Unless you come up with something better in which case I'll be happy to copy that idea from you.

whot commented on ratbag/drivers/sinowealth.py:
```diff
@@ -0,0 +1,318 @@
+        reply = Reply()
+        Parser.to_object(reply_data, self.reply_spec, obj=reply)
+        self.parse_reply(reply)
+        return reply
```
I've played with this a bit in the `hidpp20` driver and found a nice solution for debugging, see 5d6beec22de47188590122463d48be6a94db85e8.

You can pass a class name (or type) into `to_object` and it'll create an `attr` class with that name which has a reasonable default `__str__`. This may be interesting here, see the first hunk of the linked commit.

staticssleever668 replied:
I don't understand, would something change this way?

Is this related in any way with `Cannot access member "version" for type "Reply"    Member "<anything>" is unknown` warnings? (these are Pyright warnings, but mypy can do them too if you use `--strict` argument). 

Edit: type annotating `parse_reply` made this warning appear even without even the argument, hmm.

whot replied:
the main advantage is that you get a better debugging output since it's no longer a generic `Reply` object but a named object (with `attr` attributes).

I'm not sure how to get rid of this warning though because all fields on the object are dynamic, so mypy/pyright can't check anything here. I'm currently experimenting with [this approach](https://github.com/whot/ratbag-python/commit/049fd7deebcb62f6df6377479fa13308b5a42594) in the hidpp20 driver for LEDs. Could be used for query/reply handling this way too and would help with those warnings but I'm not sure it's worth it because it requires some duplication. Just as a FYI.

whot commented on ratbag/drivers/sinowealth.py:
```diff
@@ -0,0 +1,318 @@
+                Spec(
+                    "BBBB",
+                    "version",
+                    convert_from_data=lambda s: bytes(s).decode("utf-8"),
```
you can now do this as `Spec("4s", "version", convert_from_data=lambda s: s.decode("utf-8"))`, see 4b387f23430395ad9b85ac5f7e2a17311c638e4d. Takes the `int` step out of the `bytes -> ints -> bytes -> string` conversion

staticssleever668 replied:
Nice, thanks.

staticssleever668 committed:
sinowealth: use the new way for parsing strings

staticssleever668 committed:
sinowealth: add missing type hints

staticssleever668 committed:
sinowealth: fix a type error

staticssleever668 committed:
sinowealth: put ratbag.Device creation inside SinowealthDevice

staticssleever668 committed:
sinowealth: pass firmware version to the ratbag.Device

whot commented:
So, depressingly - I think I never saw the above changes. Apologies for letting this slip for that long but I think this looks good. Does it work? :)

staticssleever668 committed:
sinowealth: fix resolution creation

Broke somewhere between 0445ad36bb1748edea67b3707c208c28e5008038 and
42289d8d1f6fd1a25a1701f5d9bcce8ca3221a85, although I don't see any
related changes.

staticssleever668 commented:
Rebased.

@whot, sorry, I haven't taken a look a this in a very long time.
To be honest, I don't really have any motivation to work on this.
Also, since then the sinowealth driver in libratbag has changed a lot.
Still, I'll take a look at it now, although I don't think I will get it into an acceptable state.

Right now, reading from the mouse does seem to work, but the max DPI is incorrect (because the sensor type is not read from the mouse) and every resolution profile is set as "enabled" even when they aren't.

EDIT: `every resolution profile is set as "enabled" even when they aren't.` - this is wrong, it works correctly. However, the current active resolution ID is not set.

staticssleever668 committed:
sinowealth: use the constructor to make ratbag.Device

staticssleever668 committed:
sinowealth: don't pass an empty capability list to ratbag.Profile

staticssleever668 committed:
sinowealth: increase log level needed to print firmware version

staticssleever668 committed:
sinowealth: add `sensor_type` byte to configuration data spec

staticssleever668 committed:
sinowealth: parse the sensor type

staticssleever668 committed:
sinowealth: support reading DPI from different sensors

staticssleever668 committed:
sinowealth: support getting max DPI values per sensor type

staticssleever668 committed:
sinowealth: note to be cautious about off-by-one errors in active_dpi

staticssleever668 committed:
sinowealth: add active resolution slot reading

staticssleever668 commented:
Added sensor type and current resolution slot reading.
Now only LED and button configurations are missing.

Regarding the failing tests: I'm not sure how to fix that, I understand `Reply` is a strongly typed object returned by `Query.run` so we have to overload Query somehow to return a different type of Reply, but don't have an idea how to actually implement that.

Also one thing: I can't test the `ReportID.CONFIG_LONG` trickery as I don't have a mouse that does that anymore. This whole thing is very error prone, so maybe we should remove it all-together until someone who has such mouse can work on it.

whot commented:
> @whot, sorry, I haven't taken a look a this in a very long time.
> To be honest, I don't really have any motivation to work on this.

yeah, I understand and I too struggled to find time for this recently. There's a bit of a chicken and egg situation with this rewrite - until we have driver support there's no motivation to work on it, but without others working on it we'll never get there :) Thanks for your efforts though.

> Regarding the failing tests: I'm not sure how to fix that, I understand Reply is a strongly typed object returned by Query.run so we have to overload Query somehow to return a different type of Reply, but don't have an idea how to actually implement that.

`Reply` is just a class that gets dynamically added attributes depending on the type of query. e.g. the `reply_spec` for `QueryFWVersion` has a `report_id`, `cmd` and `version`, so the `Reply` returned by this query will have those three properties. This way we can have the code using the `Reply` object look sane. The *type* of `Reply` is always reply though, it just uses the python ability to do this:

```python
class Foo(object):
    pass

f = Foo()
f.bar = "bar"
```

In the hidpp20 code I started separating this a bit, so the replies are actually strongly typed but this was an idea I had after this driver - and tbh I think this driver isn't as complicated as HID++ so it doesn't need it as much.

This obviously doesn't play well with mypy but at least here the best solution is to just use `#  type: ignore` in those lines.
